hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/scm/client/OMBlockPrefetchClient.java

```diff
@@ -0,0 +1,369 @@
+        }
+
+        int queueSize = queue.size();
+        if (queueSize < minBlocks) {
+          int blocksToPrefetch = maxBlocks - queueSize;
+          LOG.debug(
+              "Cache for {} is below threshold (size: {}, min: {}). Submitting async prefetch task for {} blocks.",
+              replicationConfig, queueSize, minBlocks, blocksToPrefetch);
+          submitPrefetchTask(scmBlockSize, blocksToPrefetch, replicationConfig, serviceID);
+        }
+
+        metrics.addReadFromQueueLatency(Time.monotonicNowNanos() - readStartTime);
+        return allocatedBlocks;
+
+      } else {
+        LOG.debug("Bypassing cache for {}. Reason: {}", replicationConfig, queue == null ?
+            "Unsupported replication config for caching." : "ExcludeList provided.");
+        metrics.addReadFromQueueLatency(Time.monotonicNowNanos() - readStartTime);
+        metrics.incrementCacheMisses();
+        return scmBlockLocationProtocol.allocateBlock(scmBlockSize, numBlocks, replicationConfig, serviceID,
+            excludeList, clientMachine);
+      }
+    } else {
+      return scmBlockLocationProtocol.allocateBlock(scmBlockSize, numBlocks, replicationConfig, serviceID, excludeList,
+          clientMachine);
+    }
+  }
+
+  private void submitPrefetchTask(long blockSize, int blocksToPrefetch, ReplicationConfig repConfig, String serviceID) {
+
+    if (!isPrefetching.compareAndSet(false, true)) {
+      LOG.debug("Prefetch already in progress. Skipping new task for {}.", repConfig);
+      return;
+    }
+
+    if (prefetchExecutor == null || prefetchExecutor.isShutdown()) {
+      LOG.warn("Async prefetch executor is not running or shutdown. Skipping prefetch task for {}.", repConfig);
+      return;
+    }
+
+    prefetchExecutor.submit(() -> {
+      try {
+        List<AllocatedBlock> prefetchedBlocks = captureLatencyNs(metrics.getPrefetchLatencyNs(),
+            () -> scmBlockLocationProtocol.allocateBlock(blockSize, blocksToPrefetch,
+            repConfig, serviceID, new ExcludeList(), null));
+        if (prefetchedBlocks != null && !prefetchedBlocks.isEmpty()) {
+          ConcurrentLinkedDeque<ExpiringAllocatedBlock> queue = blockQueueMap.get(repConfig);
+          if (queue != null) {
+            long expiryTime = System.currentTimeMillis() + expiryDuration;
+            int addedCount = 0;
+            for (AllocatedBlock block : prefetchedBlocks) {
+              if (queue.size() < maxBlocks) {
+                queue.offer(new ExpiringAllocatedBlock(block, expiryTime));
+                addedCount++;
+              }
+            }
+            LOG.debug("Prefetched {} blocks for replication config {}.", addedCount, repConfig);
+          }
+        }
+      } catch (IOException e) {
+        LOG.error("Exception occurred while prefetching blocks.", e);
+      } finally {
+        isPrefetching.set(false);
+      }
+    });
+  }
+
+  public List<DatanodeDetails> sortDatanodes(List<DatanodeDetails> nodes, String clientMachine,
```

Comment:
I think sort datanode is also done for read request, can both use same utility class and instance ?

Reply:
Thank you, made the required changes so that the same instance can be used at both places!